services/vm-watcher/vm_watcher_service.go

```diff
@@ -132,6 +132,17 @@ func (s Service) vmFolderMoRefWithIDs(
 		z := zones.Items[i]
 
 		if v := z.Spec.ManagedVMs.FolderMoID; v != "" {
+
+			// If a zone is being deleted and it does not have any
+			// finalizer, the zone controller has already stopped the
+			// watcher and removed the finalizer. Or, a zone is being
+			// deleted even before a watcher could start. In either
+			// case, no need to watch this folder.
+			if !z.DeletionTimestamp.IsZero() &&
+				!controllerutil.ContainsFinalizer(&z, zonectrl.Finalizer) {
+				continue
```

Comment:
Same nit I give to others, by avoiding the use of continue you can avoid having to produce coverage that can be difficult to cause on purpose. In this case though I think you are able to do it without much effort.